### PR TITLE
Speed up hashing-dependent features like caching or selective testing

### PR DESCRIPTION
**Pull request by pepicrft**
Resolves https://github.com/tuist/tuist/issues/7018
XcodeGraph PR https://github.com/tuist/XcodeGraph/pull/74

### Short description 📝
This PR prevents [this scenario](https://github.com/tuist/tuist/issues/7018) from happening and speeds up the performance of hashing since we can obtain the hash of external targets using the resolved reference of the package they represent.

### How to test the changes locally 🧐
Reproducing [this](https://github.com/tuist/tuist/issues/7018) is a bit tricky since it requires creating a release version and running it from iTerm or the macOS terminal. 

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [ ] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Tests/TuistHasherTests/TargetContentHasherTests.swift [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    var sourceFilesContentHasher: MockSourceFilesContentHashing!
+    var targetScriptsContentHasher: MockTargetScriptsContentHashing!
+    var resourcesContentHasher: MockResourcesContentHashing!
+    var copyFilesContentHasher: MockCopyFilesContentHashing!
+    var headersContentHasher: MockHeadersContentHashing!
+    var deploymentTargetContentHasher: MockDeploymentTargetsContentHashing!
+    var plistContentHasher: MockPlistContentHashing!
+    var settingsContentHasher: MockSettingsContentHashing!
+    var dependenciesContentHasher: MockDependenciesContentHashing!
+    var subject: TargetContentHasher!
```
Comment (fortmarek) [inline Tests/TuistHasherTests/TargetContentHasherTests.swift:23]: nit: I'd mark all of these `private` to align with other test cases.

File: Tests/TuistHasherTests/TargetContentHasherTests.swift [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        copyFilesContentHasher = nil
+        headersContentHasher = nil
+        plistContentHasher = nil
+        settingsContentHasher = nil
+        dependenciesContentHasher = nil
+        subject = nil
+        try await super.tearDown()
+    }
+
+    func test_hash_when_targetBelongsToExternalProjectWithHash() async throws {
```
Comment (fortmarek) [inline Tests/TuistHasherTests/TargetContentHasherTests.swift:68]: I'd add a test case for when the project is _not_ external and when the project _is_ external but the hash is missing

**Review by fortmarek — DISMISSED** [review]
I was curious what the performance impact is:

The `4.34.0` release numbers for a real-world project:
```
Benchmark 1: tuist generate --no-open
  Time (mean ± σ):      4.361 s ±  0.053 s    [User: 9.011 s, System: 4.202 s]
  Range (min … max):    4.290 s …  4.455 s    10 runs
```

The performance measurements with the changes from this branch:
```
Benchmark 1: "/Users/marekfort/Library/Developer/Xcode/DerivedData/TuistExt-fnvyzgrgowotpwgcllhszvqbqzxi/Build/Products/Release/tuist_ext" generate --no-open
  Time (mean ± σ):      3.726 s ±  0.201 s    [User: 7.632 s, System: 2.715 s]
  Range (min … max):    3.553 s …  4.268 s    10 runs
```

Soo, overall ~15 %, not bad, I'd say!

File: Tests/TuistHasherTests/TargetContentHasherTests.swift (hunk 2 above)
Comment (pepicrft) [reply Tests/TuistHasherTests/TargetContentHasherTests.swift:68]: I'll open another PR with those so that we can merge this one and get the release out.

**Pull request merged**